### PR TITLE
Add protocol fee

### PR DESCRIPTION
**Pull request by Orland0x**

### CODE REVIEW

**Review by bonustrack — COMMENTED** [review]
@Orland0x About the token fee wdyt if we could go with the other approach in which the fee stay in the Boost contract and make it clear we dont support rebase tokens?

**Comment by Orland0x** [comment]
> @Orland0x About the token fee wdyt if we could go with the other approach in which the fee stay in the Boost contract and make it clear we dont support rebase tokens?

yeah we could do that. The collect token fee function would be efficient if lots of the same token gets deposited, however if all the deposits are for different tokens then collection will be pretty expensive (1 erc20 transfer for each token)  and will be paid by us rather than the DAO using it

**Comment by Orland0x** [comment]
> > @Orland0x About the token fee wdyt if we could go with the other approach in which the fee stay in the Boost contract and make it clear we dont support rebase tokens?
> 
> yeah we could do that. The collect token fee function would be efficient if lots of the same token gets deposited, however if all the deposits are for different tokens then collection will be pretty expensive (1 erc20 transfer for each token) and will be paid by us rather than the DAO using it

Maybe this is worthwhile. In the worst case it will be like we are subsiding the protocol which is probably a good thing in the beginning. Can always update if it gets too expensive later on

File: src/Boost.sol [hunk 1]
```diff
@@ -1,87 +1,145 @@
...
         keccak256("Claim(uint256 boostId,address recipient,uint256 amount)");
 
     uint256 public nextBoostId = 1;
     mapping(uint256 => BoostConfig) public boosts;
     mapping(address => mapping(uint256 => bool)) public claimed;
 
+    // Constant eth fee (in gwei) that is the same for all boost creators.
+    uint256 public ethFee;
+    // The fraction of the total boost deposit that is taken as a fee.
+    // represented as an integer denominator (1/x)%
```
Comment (pscott) [inline src/Boost.sol:23]: it's not a `%` though. If I put 2, I will get 1/x -> 1/2 -> 0.5. If I want a percentage, I need to multiply by 100! :)

File: src/Boost.sol [hunk 2]
```diff
@@ -1,87 +1,145 @@
...
+    ) external payable override {
+        if (_amount == 0) revert BoostDepositRequired();
+        if (_end <= block.timestamp) revert BoostEndDateInPast();
+        if (_start >= _end) revert BoostEndDateBeforeStart();
+        if (_guard == address(0)) revert InvalidGuard();
+        if (msg.value < ethFee) revert InsufficientEthFee();
+
+        uint256 balance = 0;
+        if (tokenFee > 0) {
+            uint256 tokenFeeAmount = _amount / tokenFee;
```
Comment (pscott) [inline src/Boost.sol:72]: This brings the question: should we also disallow `tokenFee = 1`? Creating a boost with `0` makes no sense I think?

**Review by pscott — COMMENTED** [review]
I think the tests could use some spacing (to make them more readable) as well as some light comments in order to quickly understand what's happening

File: src/Boost.sol (hunk 1 above)
Comment (Orland0x) [reply src/Boost.sol:23]: yep true, ill update the comment. i just copied uniswap: https://github.com/Uniswap/v3-core/blob/05c10bf6d547d6121622ac51c457f93775e1df09/contracts/UniswapV3Pool.sol#L68

File: src/Boost.sol (hunk 2 above)
Comment (Orland0x) [reply src/Boost.sol:72]: the protocol owner sets the fee so they can just not set it to 1. do you think we should actually prevent it?

Comment (pscott) [reply src/Boost.sol:72]: yeah just wondering why would anyone want to use boost if `tokenFee == 1`

**Pull request merged**